### ISSUE TITLE
Set global default font color

### ISSUE THREAD

**Issue by ShatReal** [godotengine/godot-proposals#1516]
**Describe the project you are working on:**
A project that has buttons and backgrounds with light colors.

**Describe the problem or limitation you are having in your project:**
The default white text isn't very visible on the backgrounds. And I have to change all of the colors individually in the theme editor to get them to show up.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
An option in project settings to set the default font color for all text would save me some time.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
Under GUI --> Theme in Project Settings how about an option for "Custom Font Color"?

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
No. I would use it often, at least.

**Is there a reason why this should be core and not an add-on in the asset library?:**
Being able to set the font color once for the entire project sounds pretty core to me.

**Comment by Calinou** [comment]
You can set a custom theme already. Shouldn't we improve the theme editor instead to make it faster to override font colors?

> Being able to set the font color once for the entire project sounds pretty core to me.

In practice, you will rarely use a single font color for everything, especially in a polished game.

**Comment by ShatReal** [comment]
Yeah, I guess so. The way you add new font colors is pretty tedious.

**Comment by winston-yallow** [comment]
> In practice, you will rarely use a single font color for everything, especially in a polished game.
In games maybe not so much, but I see more and more people use Godot for non-game applications (like a astronomical data visualizer I've seen recently). In these you most often have one primary UI font color, with maybe a few rare exceptions.

I personally have needed this feature often enough that I made my own small script that takes a theme and sets all font colors. It's a bit hacky and not really nice to use, but it gets the job done. Would love if some form of support for this were built in, either via an improvement to the theme editor or via a global default color. I personally have a slight preference for adding this to the theme editor.

**Comment by Zireael07** [comment]
> (like a astronomical data visualizer I've seen recently).
Offtop: can I have a link? pleeeease...

**Comment by winston-yallow** [comment]
> Offtop: can I have a link? pleeeease...

Not really, it's still in development by a fellow student at my university

**Comment by yankscally** [comment]
![image](https://github.com/godotengine/godot-proposals/assets/27631913/ee30402f-5ba4-4e9a-91dd-edc7e8cbb0a6)

I ran into this problem personally making a simple project, as the above user @winston-yallow stated, it's a non-game project.

I set my `rendering/environment/defaults/default_clear_color` to pure white.

Godot's default theme makes every text element white, or greyish.

My options are:

- use a script to do this, which is fine. (at runtime)
- use a signal, font override colour for every element inside a a set node (also at runtime)
- manually set each font colour in the theme, which is not ideal. 


But what if I want to change my background to any colour? now the problem is that if I change my background to the same colour as the font, there is no easy way for this to be dynamically inverted so its readable.

Unless I'm missing something?

**Comment by YuriSizov** [comment]
Changing properties in your theme at runtime to match the background color seems like the best available solution. Unfortunately, font colors are not a special property, it's just a color defined for a type of control that happens to be used for fonts. If you want to change it globally you need to change it with a theme.

Even if we were to somehow introduce a fallback font color property, there is no way to unset the color already defined for control types, so that fallback value would be unreachable. Another option would be to somehow enforce a global color override that would ignore all themes entirely (but perhaps would respect local overrides for individual controls?) But at that point changing your theme colors or injecting an intermediate theme would be a better option, a more versatile option, a more predictable option.

**Comment by SirLich** [comment]
The theme has `Default Font Size`. Could we not use the same system to introduce `Default Font Color`?

**Comment by YuriSizov** [comment]
@SirLich Font sizes can be set to 0 to imply "no value", but what can colors be set to to imply the same?

**Comment by SirLich** [comment]
> The default font size of this theme resource. Used as the default value when trying to fetch a font size value that doesn't exist in this theme or is in invalid state. If the default font size is also missing or invalid, the engine fallback value is used (see ThemeDB.fallback_font_size).

> Values below 1 are invalid and can be used to unset the property. Use has_default_font_size() to check if this value is valid.

My understanding of themes is that the first part ("... that doesn't exist in this theme") is easy enough to do. So if the only consideration is the second part ("Values below 1 are invalid and can be used to unset the property"), then that's easily solvable by having a secondary flag. 

We can even use the same pattern that property overides themselves do (namely, put the gate inline): 

<img width="602" height="146" alt="Image" src="https://github.com/user-attachments/assets/0fa7d390-f72d-463e-914d-961147c77235" />

**Comment by YuriSizov** [comment]
I was rather thinking about individual color properties, but I guess you would just not add them to your theme if you don't want them to be valid. Font sizes, in contrast, can be added but set to 0 to make them invalid. For the default theme value specifically, yes, a checkbox is possible.

That's only a part of the problem though. The other is how to associate specific colors with being a "font color" and resolve them differently from any other color.

**Comment by toupeira** [comment]
Not sure if this makes sense, but maybe the base `Control` node could be added to the theme editor somehow? So overriding e.g. `font_color` there would cause all other controls to inherit it.

**Comment by YuriSizov** [comment]
@toupeira There is no special meaning behind the color named `font_color`. It's just a property name, for all intents and purposes. And some controls may use a different name for it, or have multiple colors related to rendering text. There is no specialized "font color" theme item kind, only "color". And to support the proposal here the engine would need to introduce a special way to defining and resolving colors specifically related to text rendering, and that gets pretty convoluted.

I mean, we did get font size items, separate from integer constants that we had before. So we could add special font colors too, but I'm not particularly fond of this direction.
